File: packages/protocol/src/crypto.ts
import { chacha20poly1305, xchacha20poly1305 } from "@noble/ciphers/chacha.js";
import { x25519 } from "@noble/curves/ed25519.js";
import { hkdf } from "@noble/hashes/hkdf.js";
import { hmac } from "@noble/hashes/hmac.js";
import { sha256 } from "@noble/hashes/sha2.js";
import { bytesToHex, concatBytes, hexToBytes, randomBytes, utf8ToBytes } from "@noble/hashes/utils.js";

const INFO_PAIR = utf8ToBytes("zimlo-pair-v1");
const INFO_DEVICE = utf8ToBytes("zimlo-device-v1");
const INFO_CLIENT_TX = utf8ToBytes("zimlo-ws-client-tx-v1");
const INFO_SERVER_TX = utf8ToBytes("zimlo-ws-server-tx-v1");
const INFO_PUSH_ROUTE = utf8ToBytes("zimlo-push-route-v1");

export function toBase64Url(bytes: Uint8Array): string {
  let binary = "";
  for (const byte of bytes) binary += String.fromCharCode(byte);
  return btoa(binary).replaceAll("+", "-").replaceAll("/", "_").replace(/=+$/u, "");
}

export function fromBase64Url(value: string): Uint8Array {
  const normalized = value.replaceAll("-", "+").replaceAll("_", "/");
  const padded = normalized.padEnd(Math.ceil(normalized.length / 4) * 4, "=");
  const binary = atob(padded);
  return Uint8Array.from(binary, (character) => character.charCodeAt(0));
}

export function createKeyPair(): { privateKey: Uint8Array; publicKey: Uint8Array } {
  const privateKey = x25519.utils.randomSecretKey();
  return { privateKey, publicKey: x25519.getPublicKey(privateKey) };
}

export function derivePairKey(
  privateKey: Uint8Array,
  peerPublicKey: Uint8Array,
  secret: Uint8Array,
): Uint8Array {
  return hkdf(sha256, x25519.getSharedSecret(privateKey, peerPublicKey), secret, INFO_PAIR, 32);
}

export function deriveDeviceKey(pairKey: Uint8Array, secret: Uint8Array): Uint8Array {
  return hkdf(sha256, pairKey, secret, INFO_DEVICE, 32);
}

export function makeProof(key: Uint8Array, message: string): string {
  return toBase64Url(hmac(sha256, key, utf8ToBytes(message)));
}

export function verifyProof(key: Uint8Array, message: string, proof: string): boolean {
  const actual = makeProof(key, message);
  if (actual.length !== proof.length) return false;
  let mismatch = 0;
  for (let index = 0; index < actual.length; index += 1) {
    mismatch |= actual.charCodeAt(index) ^ proof.charCodeAt(index);
  }
  return mismatch === 0;
}

export function deriveConnectionKeys(
  deviceKey: Uint8Array,
  clientNonce: Uint8Array,
  serverNonce: Uint8Array,
): { clientTx: Uint8Array; serverTx: Uint8Array } {
  const salt = concatBytes(clientNonce, serverNonce);
  return {
    clientTx: hkdf(sha256, deviceKey, salt, INFO_CLIENT_TX, 32),
    serverTx: hkdf(sha256, deviceKey, salt, INFO_SERVER_TX, 32),
  };
}

function counterNonce(counter: number): Uint8Array {
  const nonce = new Uint8Array(24);
  const view = new DataView(nonce.buffer);
  view.setBigUint64(16, BigInt(counter), false);
  return nonce;
}

export function encryptFrame(key: Uint8Array, counter: number, value: unknown, aad: string): string {
  const plaintext = utf8ToBytes(JSON.stringify(value));
  return toBase64Url(xchacha20poly1305(key, counterNonce(counter), utf8ToBytes(aad)).encrypt(plaintext));
}

export function decryptFrame<T>(key: Uint8Array, counter: number, ciphertext: string, aad: string): T {
  const plaintext = xchacha20poly1305(key, counterNonce(counter), utf8ToBytes(aad)).decrypt(
    fromBase64Url(ciphertext),
  );
  return JSON.parse(new TextDecoder().decode(plaintext)) as T;
}

export interface PushRouteEnvelope {
  ephemeralPublicKey: string;
  nonce: string;
  ciphertext: string;
}

export function sealPushRoute(peerPublicKey: Uint8Array, value: unknown): PushRouteEnvelope {
  const ephemeralPrivateKey = x25519.utils.randomSecretKey();
  const ephemeralPublicKey = x25519.getPublicKey(ephemeralPrivateKey);
  const shared = x25519.getSharedSecret(ephemeralPrivateKey, peerPublicKey);
  const key = hkdf(sha256, shared, new Uint8Array(), INFO_PUSH_ROUTE, 32);
  const nonce = randomBytes(12);
  const plaintext = utf8ToBytes(JSON.stringify(value));
  const ciphertext = chacha20poly1305(key, nonce, INFO_PUSH_ROUTE).encrypt(plaintext);
  return {
    ephemeralPublicKey: toBase64Url(ephemeralPublicKey),
    nonce: toBase64Url(nonce),
    ciphertext: toBase64Url(ciphertext),
  };
}

export function openPushRoute<T>(privateKey: Uint8Array, envelope: PushRouteEnvelope): T {
  const shared = x25519.getSharedSecret(privateKey, fromBase64Url(envelope.ephemeralPublicKey));
  const key = hkdf(sha256, shared, new Uint8Array(), INFO_PUSH_ROUTE, 32);
  const plaintext = chacha20poly1305(
    key,
    fromBase64Url(envelope.nonce),
    INFO_PUSH_ROUTE,
  ).decrypt(fromBase64Url(envelope.ciphertext));
  return JSON.parse(new TextDecoder().decode(plaintext)) as T;
}

export { bytesToHex, concatBytes, hexToBytes, randomBytes };

File: runtime/crates/zimlo-protocol/src/crypto.rs
use base64::{Engine as _, engine::general_purpose::URL_SAFE_NO_PAD};
use chacha20poly1305::{
    ChaCha20Poly1305, XChaCha20Poly1305,
    aead::{Aead, KeyInit, Payload},
};
use hkdf::Hkdf;
use hmac::{Hmac, Mac};
use serde::{Deserialize, Serialize, de::DeserializeOwned};
use sha2::Sha256;
use thiserror::Error;
use x25519_dalek::{X25519_BASEPOINT_BYTES, x25519};

const INFO_PAIR: &[u8] = b"zimlo-pair-v1";
const INFO_DEVICE: &[u8] = b"zimlo-device-v1";
const INFO_CLIENT_TX: &[u8] = b"zimlo-ws-client-tx-v1";
const INFO_SERVER_TX: &[u8] = b"zimlo-ws-server-tx-v1";
const INFO_PUSH_ROUTE: &[u8] = b"zimlo-push-route-v1";

#[derive(Debug, Error)]
pub enum CryptoError {
    #[error("invalid {name} length: expected {expected}, got {actual}")]
    InvalidLength {
        name: &'static str,
        expected: usize,
        actual: usize,
    },
    #[error("invalid base64url value")]
    InvalidBase64(#[from] base64::DecodeError),
    #[error("key derivation failed")]
    KeyDerivation,
    #[error("authenticated encryption failed")]
    Encryption,
    #[error("authenticated decryption failed")]
    Decryption,
    #[error("invalid authentication proof")]
    InvalidProof,
    #[error("random byte generation failed")]
    Random(#[from] getrandom::Error),
    #[error("JSON encoding failed")]
    Json(#[from] serde_json::Error),
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct KeyPair {
    pub private_key: [u8; 32],
    pub public_key: [u8; 32],
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ConnectionKeys {
    pub client_tx: [u8; 32],
    pub server_tx: [u8; 32],
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PushRouteEnvelope {
    pub ephemeral_public_key: String,
    pub nonce: String,
    pub ciphertext: String,
}

pub fn to_base64_url(bytes: &[u8]) -> String {
    URL_SAFE_NO_PAD.encode(bytes)
}

pub fn from_base64_url(value: &str) -> Result<Vec<u8>, CryptoError> {
    Ok(URL_SAFE_NO_PAD.decode(value)?)
}

pub fn fixed_bytes<const N: usize>(
    name: &'static str,
    bytes: &[u8],
) -> Result<[u8; N], CryptoError> {
    bytes.try_into().map_err(|_| CryptoError::InvalidLength {
        name,
        expected: N,
        actual: bytes.len(),
    })
}

pub fn create_key_pair() -> Result<KeyPair, CryptoError> {
    let private_key = random_bytes()?;
    Ok(KeyPair {
        public_key: public_key(&private_key),
        private_key,
    })
}

pub fn random_bytes<const N: usize>() -> Result<[u8; N], CryptoError> {
    let mut bytes = [0_u8; N];
    getrandom::fill(&mut bytes)?;
    Ok(bytes)
}

pub fn public_key(private_key: &[u8; 32]) -> [u8; 32] {
    x25519(*private_key, X25519_BASEPOINT_BYTES)
}

pub fn derive_pair_key(
    private_key: &[u8; 32],
    peer_public_key: &[u8; 32],
    secret: &[u8],
) -> Result<[u8; 32], CryptoError> {
    derive_key(&x25519(*private_key, *peer_public_key), secret, INFO_PAIR)
}

pub fn derive_device_key(pair_key: &[u8; 32], secret: &[u8]) -> Result<[u8; 32], CryptoError> {
    derive_key(pair_key, secret, INFO_DEVICE)
}

pub fn make_proof(key: &[u8], message: &str) -> Result<String, CryptoError> {
    let mut mac =
        <Hmac<Sha256> as Mac>::new_from_slice(key).map_err(|_| CryptoError::InvalidProof)?;
    mac.update(message.as_bytes());
    Ok(to_base64_url(&mac.finalize().into_bytes()))
}

pub fn verify_proof(key: &[u8], message: &str, proof: &str) -> Result<(), CryptoError> {
    let proof = from_base64_url(proof)?;
    let mut mac =
        <Hmac<Sha256> as Mac>::new_from_slice(key).map_err(|_| CryptoError::InvalidProof)?;
    mac.update(message.as_bytes());
    mac.verify_slice(&proof)
        .map_err(|_| CryptoError::InvalidProof)
}

pub fn derive_connection_keys(
    device_key: &[u8; 32],
    client_nonce: &[u8; 24],
    server_nonce: &[u8; 24],
) -> Result<ConnectionKeys, CryptoError> {
    let mut salt = [0_u8; 48];
    salt[..24].copy_from_slice(client_nonce);
    salt[24..].copy_from_slice(server_nonce);
    Ok(ConnectionKeys {
        client_tx: derive_key(device_key, &salt, INFO_CLIENT_TX)?,
        server_tx: derive_key(device_key, &salt, INFO_SERVER_TX)?,
    })
}

pub fn encrypt_frame<T: Serialize>(
    key: &[u8; 32],
    counter: u64,
    value: &T,
    aad: &str,
) -> Result<String, CryptoError> {
    let cipher = XChaCha20Poly1305::new_from_slice(key).map_err(|_| CryptoError::Encryption)?;
    let plaintext = serde_json::to_vec(value)?;
    let ciphertext = cipher
        .encrypt(
            (&counter_nonce(counter)).into(),
            Payload {
                msg: &plaintext,
                aad: aad.as_bytes(),
            },
        )
        .map_err(|_| CryptoError::Encryption)?;
    Ok(to_base64_url(&ciphertext))
}

pub fn decrypt_frame<T: DeserializeOwned>(
    key: &[u8; 32],
    counter: u64,
    ciphertext: &str,
    aad: &str,
) -> Result<T, CryptoError> {
    let cipher = XChaCha20Poly1305::new_from_slice(key).map_err(|_| CryptoError::Decryption)?;
    let ciphertext = from_base64_url(ciphertext)?;
    let plaintext = cipher
        .decrypt(
            (&counter_nonce(counter)).into(),
            Payload {
                msg: &ciphertext,
                aad: aad.as_bytes(),
            },
        )
        .map_err(|_| CryptoError::Decryption)?;
    Ok(serde_json::from_slice(&plaintext)?)
}

pub fn seal_push_route<T: Serialize>(
    peer_public_key: &[u8; 32],
    value: &T,
) -> Result<PushRouteEnvelope, CryptoError> {
    let ephemeral = create_key_pair()?;
    let mut nonce = [0_u8; 12];
    getrandom::fill(&mut nonce)?;
    seal_push_route_with_material(peer_public_key, &ephemeral.private_key, &nonce, value)
}

pub fn seal_push_route_with_material<T: Serialize>(
    peer_public_key: &[u8; 32],
    ephemeral_private_key: &[u8; 32],
    nonce: &[u8; 12],
    value: &T,
) -> Result<PushRouteEnvelope, CryptoError> {
    let ephemeral_public_key = public_key(ephemeral_private_key);
    let shared = x25519(*ephemeral_private_key, *peer_public_key);
    let key = derive_key(&shared, &[], INFO_PUSH_ROUTE)?;
    let cipher = ChaCha20Poly1305::new_from_slice(&key).map_err(|_| CryptoError::Encryption)?;
    let plaintext = serde_json::to_vec(value)?;
    let ciphertext = cipher
        .encrypt(
            nonce.into(),
            Payload {
                msg: &plaintext,
                aad: INFO_PUSH_ROUTE,
            },
        )
        .map_err(|_| CryptoError::Encryption)?;
    Ok(PushRouteEnvelope {
        ephemeral_public_key: to_base64_url(&ephemeral_public_key),
        nonce: to_base64_url(nonce),
        ciphertext: to_base64_url(&ciphertext),
    })
}

pub fn open_push_route<T: DeserializeOwned>(
    private_key: &[u8; 32],
    envelope: &PushRouteEnvelope,
) -> Result<T, CryptoError> {
    let ephemeral_public_key = fixed_bytes::<32>(
        "ephemeral public key",
        &from_base64_url(&envelope.ephemeral_public_key)?,
    )?;
    let nonce = fixed_bytes::<12>("push route nonce", &from_base64_url(&envelope.nonce)?)?;
    let ciphertext = from_base64_url(&envelope.ciphertext)?;
    let shared = x25519(*private_key, ephemeral_public_key);
    let key = derive_key(&shared, &[], INFO_PUSH_ROUTE)?;
    let cipher = ChaCha20Poly1305::new_from_slice(&key).map_err(|_| CryptoError::Decryption)?;
    let plaintext = cipher
        .decrypt(
            (&nonce).into(),
            Payload {
                msg: &ciphertext,
                aad: INFO_PUSH_ROUTE,
            },
        )
        .map_err(|_| CryptoError::Decryption)?;
    Ok(serde_json::from_slice(&plaintext)?)
}

fn derive_key(
    input_key_material: &[u8],
    salt: &[u8],
    info: &[u8],
) -> Result<[u8; 32], CryptoError> {
    let hkdf = Hkdf::<Sha256>::new(Some(salt), input_key_material);
    let mut output = [0_u8; 32];
    hkdf.expand(info, &mut output)
        .map_err(|_| CryptoError::KeyDerivation)?;
    Ok(output)
}

fn counter_nonce(counter: u64) -> [u8; 24] {
    let mut nonce = [0_u8; 24];
    nonce[16..].copy_from_slice(&counter.to_be_bytes());
    nonce
}
